src/core/test-unit/Diffing/Internal/StateChangeFinderWorkerTests/StateChangeFinderWorker_BasicTests.cs

```diff
@@ -0,0 +1,268 @@
+            Assert.Equal("Int", change.Breadcrumb);
+            Assert.Equal(666, change.OldValue);
+            Assert.Equal(777, change.NewValue);
+            
+            change = sorted[9];
+            Assert.Equal("PrivateIsIncluded", change.Breadcrumb);
+            Assert.Equal("private", change.OldValue);
+            Assert.Equal("changed", change.NewValue);
+            
+            change = sorted[10];
+            Assert.Equal("SomeThing.SomeString", change.Breadcrumb);
+            Assert.Equal("Some", change.OldValue);
+            Assert.Equal("Changed", change.NewValue);
+            
+            change = sorted[11];
+            Assert.Equal("String", change.Breadcrumb);
+            Assert.Equal("Kewl string", change.OldValue);
+            Assert.Equal("Hot string", change.NewValue);
+            
+            change = sorted[12];
+            Assert.Equal("TimeSpan", change.Breadcrumb);
+            Assert.Equal(oldState.TimeSpan, change.OldValue);
+            Assert.Equal(newState.TimeSpan, change.NewValue);
+        }
+        
+        [Fact]
+        public void FindDifferences_recursionDepthHandling_bailButDoNotBlow()
+        {
+            var oldChild = new ChildVictim {ChildString = "old child"};
+            var oldState = new ParentVictim
+            {
+                Child = oldChild,
+                ParentString = "old parent"
+            };
+            oldChild.Parent = oldState;
+            
+            var newChild = new ChildVictim {ChildString = "new child"};
+            var newState = new ParentVictim
+            {
+                Child = newChild,
+                ParentString = "new parent"
+            };
+            newChild.Parent = newState;
+
+            worker.FindDifferences(oldState, newState, stateChanges, 3);
+            
+            Assert.Equal(4, stateChanges.Changes.Count);
+
+            var change = stateChanges.Changes[0];
+            Assert.Equal("Child.Parent.Child.ChildString", change.Breadcrumb);
+            Assert.Equal("old child", change.OldValue);
+            Assert.Equal("new child", change.NewValue);
+
+            change = stateChanges.Changes[1];
+            Assert.Equal("Child.Parent.ParentString", change.Breadcrumb);
+            Assert.Equal("old parent", change.OldValue);
+            Assert.Equal("new parent", change.NewValue);
+            
+            change = stateChanges.Changes[2];
+            Assert.Equal("Child.ChildString", change.Breadcrumb);
+            Assert.Equal("old child", change.OldValue);
+            Assert.Equal("new child", change.NewValue);
+            
+            change = stateChanges.Changes[3];
+            Assert.Equal("ParentString", change.Breadcrumb);
+            Assert.Equal("old parent", change.OldValue);
+            Assert.Equal("new parent", change.NewValue);
+        }
+
+        #region Victims
+        public class TheVictim
+        {
+            private string PrivateIsIncluded { get; set; }
```

Comment:
Is it worth testing with protected and internal members?

Reply:
Yes, otherwise someone might change the binding flags used in the reflection and then it stops working.

For the specific case of auditing, we do not need the private fields, but the StateChangeFinder is supposed to support all kinds of diffing and if you are just using it in a debugging-ish context, then those private fields etc. could be important. And they are easy to ignore in the listener.

Reply:
yes I get that, so I think my point was to actually add properties that are `protected` and `internal` modified :)

Reply:
I misunderstood :)